### PR TITLE
Add MiniMapChanged Callins

### PR DESCRIPTION
**Pull request by TheFutureKnight**
### Work Done

- Added `MiniMapRotationChanged` callin
- Added `MiniMapGeometryChanged` callin
- Added `MiniMapMinimizationChanged` callin

### Related issues

- #1994

### CODE REVIEW

**Comment by sprunk** [comment]
Looks reasonable, might want to test some edge cases (e.g. resize/minimize the game window so that the minimap gets adjusted too and see what happens).

**Comment by TheFutureKnight** [comment]
Will do, also I assume this is self explanatory from the code but 2 notes I would like to add:

- I added in minimap.h 2 new vars named lastDim and lastPos, even tho oldPos and oldDim already exists but is used by toggle maximize to stash pos and dim until the minimap no longer needs to be maximized, I assume if I re-use these vars it would clash somewhere in an edge case even tho I haven't tested it, I might test and see what happens if so I will remove the newer vars.

- I intended to use `SetGeo()` (even tho it's only used once, in gl lua) all over the minimap.cpp file instead of putting the eventhandler call inside Update() - this would probably need some time so maybe if I can find time I will see what can be done.

File: rts/Game/UI/MiniMap.cpp [hunk 1]
```diff
@@ -263,7 +263,20 @@ void CMiniMap::SetRotation(RotationOptions state) // 0 1 2 3: 0 90 180 270
     if (state == rotation)
         return;
 
+	const float oldRotation = static_cast<int>(rotation) * math::HALFPI;
     rotation = state;
+	eventHandler.MiniMapRotationChanged(static_cast<int>(rotation) * math::HALFPI, oldRotation);
```
Comment (badosu) [inline rts/Game/UI/MiniMap.cpp:268]: Aren't we doing `rotation * halfpi * halfpi` here? Also formatting

Comment (badosu) [inline rts/Game/UI/MiniMap.cpp:268]: Nitpick: formatting

File: rts/Game/UI/MiniMap.cpp [hunk 2]
```diff
@@ -1088,6 +1103,12 @@ void CMiniMap::Update()
 	fbo.Bind();
 	UpdateTextureCache();
 
+	if (curPos != lastPos || curDim != lastDim) { // probably can be moved to SetGeometry
```
Comment (badosu) [inline rts/Game/UI/MiniMap.cpp:1106]: Can we move this to `SetGeometry` or is there some concern with geometry state being changed in ways we can't track?

File: rts/Lua/LuaHandle.h [hunk 3]
```diff
@@ -214,6 +214,10 @@ class CLuaHandle : public CEventClient
 		void CameraRotationChanged(const float3& rot) override;
 		void CameraPositionChanged(const float3& pos) override;
 
+		void MiniMapRotationChanged(const float newRot, const float oldRot) override;
+		void MiniMapMinimizationChanged(const bool isMinimized) override;
```
Comment (badosu) [inline rts/Lua/LuaHandle.h:218]: Very nitpicky: I don't like the name but I can't suggest a better one either.

**Comment by badosu** [comment]
Overall LGTM, sent a few comments.

File: rts/Game/UI/MiniMap.cpp (hunk 2 above)
Comment (TheFutureKnight) [reply rts/Game/UI/MiniMap.cpp:1106]: The entire file continuously changes dim and pos from curDim and curPos regularly from many functions, a simple swap and replace doesn't work either cuz some lines require the actual curX

TLDR: still seeing if we can do so

Comment (badosu) [reply rts/Game/UI/MiniMap.cpp:1106]: Sounds good! I'll mark this as resolved as I can't offer an obvious straightforward solution that does not make the callin spammy (invoked multiple times for situations it should be condensed in a single one)

Comment (sprunk) [reply rts/Game/UI/MiniMap.cpp:1106]: If it doesn't work out consider
```suggestion
	/* Would need a larger refactor for this to become SetGeometry since the entire file
	 * continuously changes dim and pos from curDim and curPos regularly from many functions */
	if (curPos != lastPos || curDim != lastDim) {
```

File: rts/Lua/LuaHandle.cpp [hunk 4]
```diff
@@ -3422,6 +3422,83 @@ void CLuaHandle::CameraPositionChanged(const float3& pos)
...
+ * 
+ * @function Callins:MiniMapGeometryChanged
+ * @param newPosX number
+ * @param newPosY number
+ * @param newDimX number
+ * @param newDimY number
+ * @param oldPosX number
+ * @param oldPosY number
+ * @param oldDimX number
+ * @param oldDimY number
```
Comment (sprunk) [inline rts/Lua/LuaHandle.cpp:3477]: Should document what unit each of these values is in

File: rts/Lua/LuaHandle.h (hunk 3 above)
Comment (TheFutureKnight) [reply rts/Lua/LuaHandle.h:218]: MiniMapMinimizeChanged?

File: rts/Lua/LuaHandle.cpp (hunk 4 above)
Comment (TheFutureKnight) [reply rts/Lua/LuaHandle.cpp:3477]: Done, pixels.

File: rts/Lua/LuaHandle.h (hunk 3 above)
Comment (badosu) [reply rts/Lua/LuaHandle.h:218]: I like it better, but honestly all names are bad I guess, I'd love some word that described it better, keeping in mind that there also exist other display settings like maximized, maxspect etc

Comment (badosu) [reply rts/Lua/LuaHandle.h:218]: I'll just mark this as resolved for now, maybe in the future we'd rather have a callin for the low frequency callins on the list of settings that change minimap display in ways that setgeometry doesnt capture for example, minimized being one of them. But realistically it really doesn't make a ton of difference, games should be handling the minimap themselves, I guess if there's a game that actively uses maxspec or maximized or whatever  we'd have a better idea on how to handle this, but most of the time it's just minimized indeed.

Comment (sprunk) [reply rts/Lua/LuaHandle.h:218]: Hm, aren't there actually 3 modes so the bool paramater isn't sufficient? It can be maximized, minimized, or "normal" (I assume it can't be minimized and maximized at the same time but since there's two bools internally perhaps you could get into that state too?).
https://github.com/beyond-all-reason/spring/blob/607120074f5de92a87baf1d371f0a30b13156c12/rts/Game/UI/MiniMap.h#L59-L60

File: rts/Game/UI/MiniMap.cpp [hunk 5]
```diff
@@ -258,12 +258,25 @@ void CMiniMap::ToggleMaximized(bool _maxspect)
...
+
+void CMiniMap::SetMinimized(bool state)
+{
+	RECOIL_DETAILED_TRACY_ZONE;
+	
+	if (minimized == state)
+		return;
+
+	minimized = state;
+	eventHandler.MiniMapMinimizationChanged(minimized);
```
Comment (sprunk) [inline rts/Game/UI/MiniMap.cpp:279]: What about maximized? Shouldn't that one also get an event, or perhaps this event should have 2 params (minimized and maximized)?
https://github.com/beyond-all-reason/spring/blob/607120074f5de92a87baf1d371f0a30b13156c12/rts/Game/UI/MiniMap.h#L59-L60

Comment (TheFutureKnight) [reply rts/Game/UI/MiniMap.cpp:279]: Sure, do we prefer having 2 callins or just one with 2 params?

Comment (sprunk) [reply rts/Game/UI/MiniMap.cpp:279]: Check how each of the two modes actually behaves in combination with geometry getters etc. If neither turns out to be redundant then probably one callin with 2 params since they're not independent

Comment (sprunk) [reply rts/Game/UI/MiniMap.cpp]: I'm not sure, perhaps we can keep as-is and then we will decide based on feedback from users of `gl.SMM`.

**Comment by TheFutureKnight** [comment]
Does this require any more work?

**Comment by sprunk** [comment]
Code looks alright, would be good to provide some test widget that uses those. Otherwise it might be a matter of waiting because we're currently delayed a bit in the release cycle due to desync hunting.

**Comment by TheFutureKnight** [comment]
Sure, should it be a mockup that displays all callins and their args or an actual useful widget in game?

**Comment by sprunk** [comment]
Either works

**Comment by TheFutureKnight** [comment]
Added the mockup file, if it needs a change lmk

File: cont/examples/Widgets/minimap_changed_mock.lua [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+		author = "TheFutureKnight",
+		date = "2025-5-14",
+		license = "GNU GPL, v2 or later",
+		layer = 0,
+		enabled = true,
+	}
+end
+
+function widget:MiniMapRotationChanged(newRot, oldRot)
+	Spring.Echo("MMRC: from " .. oldRot .. " to " .. newRot)
```
Comment (badosu) [inline cont/examples/Widgets/minimap_changed_mock.lua:14]: Abbreviation too cryptic to be of any use, since its an example anyway no harm in actually changing to:

`	Spring.Echo("<MiniMap Callins Test> Minimap rotation changed from " .. oldRot .. " to " .. newRot)`

Comment (TheFutureKnight) [reply cont/examples/Widgets/minimap_changed_mock.lua:14]: Fair change, committing

**Comment by sprunk** [comment]
Looks good, I should be able to test on the weekend.

**Comment by WatchTheFort** [comment]
> Looks good, I should be able to test on the weekend.

How did the testing go?

**Comment by TheFutureKnight** [comment]
💀

**Review by sprunk — CHANGES_REQUESTED** [review]
Almost works!

Two related remaining issues are that
 * slave mode resets mini/maximized state without producing an event
 * slave mode isn't itself conveyed via any event.

Perhaps add a 3rd bool to `MiniMapStateChanged` so that the args are `isMinimized, isMaximized, isSlaved`, and then call this when slave state changes to solve both. This would probably go somewhere here https://github.com/beyond-all-reason/RecoilEngine/blob/77c61cfc481a9426ed15b9f406c53efd14ff88e2/rts/Game/UI/MiniMap.cpp#L334-L342

**Comment by TheFutureKnight** [comment]
The slave thing was known and discussed in one of those threads iirc, I will see that it triggers the callin at the very least

**Comment by TheFutureKnight** [comment]
@sprunk I think the last commit addresses the slave issue, however if there are more issues lmk

**Review by sprunk — APPROVED** [review]
Looks great, thank you!
